internals-js/src/operations.ts

```diff
@@ -1435,17 +1432,15 @@ export class FieldSelection extends Freezable<FieldSelection> {
   }
 
   filter(predicate: (selection: Selection) => boolean): FieldSelection | undefined {
-    if (!predicate(this)) {
-      return undefined;
-    }
     if (!this.selectionSet) {
-      return this;
+      return predicate(this) ? this : undefined;
     }
 
     const updatedSelectionSet = this.selectionSet.filter(predicate);
-    return this.selectionSet === updatedSelectionSet
+    const thisWithFilteredSelectionSet = this.selectionSet === updatedSelectionSet
```

Comment:
Are you sure you want === here? It seems a little strange to me that we'd do that type of equality test immediately after filtering, I sould expect some sort of deep equality check, but I'm not looking into it too deeply.

Reply:
Yeah, this code is just an optimisation. The point is to save allocating a new `FieldSelection` object (or, in the case of `SelectionSet.filter`, a `SelectionSet` object) in the case where nothing was filtered. Which is expected to be common, both because we use `filter` in case where we know filtering rarely kicks in, like in that patch with empty branches which happens but not at all in all cases, but also because the method is recursive and while filtering may happen in some branches of a selection set, it probably still won't happen in most of them).

Anyway, using deep equality doesn't make sense in term of such optimisation since we'd lose more in computing the deep equality check in general than what we'd save in allocations.